# Hongyi Tang HW1

# 1.List

1.1

In [2]:
movies = ["Spider Man", "Moonlight", "The Holiday", "Iron Man", "Slience", "Conjuring"]

1.2

In [4]:
first_three = movies[0:3]
first_three

['Spider Man', 'Moonlight', 'The Holiday']

1.3

In [14]:
new_list = movies[0:2] + movies[-2:]
new_list

['Spider Man', 'Moonlight', 'Slience', 'Conjuring']

# 2. Loops

2.1

In [17]:
numbers = []
for i in range(70,80):
    numbers.append(i)
numbers

[70, 71, 72, 73, 74, 75, 76, 77, 78, 79]

2.2

In [20]:
sum = 0 
for i in range(0,101):
    sum+=i
sum

5050

2.3

In [40]:
seats = {"Adam":33, "Ashin":12, "Inukai":45, "Tanaka":2, "Ikki":17}
for i in seats.keys():
    print(i + ": " + str(seats[i]))

Adam: 33
Ashin: 12
Inukai: 45
Tanaka: 2
Ikki: 17


# 3. Comprehesions

3.1

In [45]:
sq_numbers = [num**2 for num in range(1,11)]
sq_numbers

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

3.2

In [46]:
double_numbers = {k:k**2 for k in range(1,11)}
double_numbers

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100}

# 4. Functions and Data Transformation

4.1

In [53]:
def timeTransformation(t):
    hours = t//100
    mins = t%100/60
    return mins+hours

In [54]:
timeTransformation(345)

3.75

In [55]:
timeTransformation(1015)

10.25

# 5. Load Data

In [2]:
import pandas as pd
import numpy as np

5.1

In [3]:
data = pd.read_csv("gapminder.csv.bz2", sep="\t")
data

iso3      name iso2    region                       sub-region  \
0      ABW     Aruba   AW  Americas  Latin America and the Caribbean   
1      ABW     Aruba   AW  Americas  Latin America and the Caribbean   
2      ABW     Aruba   AW  Americas  Latin America and the Caribbean   
3      ABW     Aruba   AW  Americas  Latin America and the Caribbean   
4      ABW     Aruba   AW  Americas  Latin America and the Caribbean   
...    ...       ...  ...       ...                              ...   
13050  ZWE  Zimbabwe   ZW    Africa               Sub-Saharan Africa   
13051  ZWE  Zimbabwe   ZW    Africa               Sub-Saharan Africa   
13052  ZWE  Zimbabwe   ZW    Africa               Sub-Saharan Africa   
13053  ZWE  Zimbabwe   ZW    Africa               Sub-Saharan Africa   
13054  ZWE  Zimbabwe   ZW    Africa               Sub-Saharan Africa   

      intermediate-region    time  totalPopulation  fertilityRate  \
0               Caribbean  1960.0          54211.0          4.820   
1               Caribbean  1961.0          55438.0          4.655   
2               Caribbean  1962.0          56225.0          4.471   
3               Caribbean  1963.0          56695.0          4.271   
4               Caribbean  1964.0          57032.0          4.059   
...                   ...     ...              ...            ...   
13050      Eastern Africa  2015.0       13814629.0          3.896   
13051      Eastern Africa  2016.0       14030390.0          3.804   
13052      Eastern Africa  2017.0       14236745.0          3.707   
13053      Eastern Africa  2018.0       14439018.0          3.615   
13054      Eastern Africa  2019.0       14645468.0          3.531   

       lifeExpectancy  ...  accessElectricity  agriculturalLand  \
0              65.662  ...                NaN               NaN   
1              66.074  ...                NaN              20.0   
2              66.444  ...                NaN              20.0   
3              66.787  ...                NaN              20.0   
4              67.113  ...                NaN              20.0   
...               ...  ...                ...               ...   
13050          59.534  ...           33.70000          162000.0   
13051          60.294  ...           39.67623          162000.0   
13052          60.812  ...           40.14428          162000.0   
13053          61.195  ...           40.61636          162000.0   
13054          61.490  ...           41.08911               NaN   

       agricultureTractors  cerealProduction  fertilizerHa        co2  \
0                      NaN               NaN           NaN  11092.675   
1                      NaN               NaN           NaN  11576.719   
2                      NaN               NaN           NaN  12713.489   
3                      NaN               NaN           NaN  12178.107   
4                      NaN               NaN           NaN  11840.743   
...                    ...               ...           ...        ...   
13050                  NaN          799405.0         22.90  12317.453   
13051                  NaN          675062.0         32.50  10982.665   
13052                  NaN          895073.0         38.35        NaN   
13053                  NaN          950627.0         38.35        NaN   
13054                  NaN               NaN           NaN        NaN   

       greenhouseGases     co2_PC  pm2.5_35  battleDeaths  
0                  NaN  204.62037       NaN           NaN  
1                  NaN  208.82281       NaN           NaN  
2                  NaN  226.11808       NaN           NaN  
3                  NaN  214.80037       NaN           NaN  
4                  NaN  207.61578       NaN           NaN  
...                ...        ...       ...           ...  
13050              NaN    0.89162       0.0           NaN  
13051              NaN    0.78278       0.0           NaN  
13052              NaN        NaN       0.0           NaN  
13053              NaN        

5.2

In [4]:
data.shape

(13055, 25)

(a) there are 25 variables
(b) there are 13055 rows

In [75]:
data.head(3)

iso3   name iso2    region                       sub-region  \
0  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
1  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
2  ABW  Aruba   AW  Americas  Latin America and the Caribbean   

  intermediate-region    time  totalPopulation  fertilityRate  lifeExpectancy  \
0           Caribbean  1960.0          54211.0          4.820          65.662   
1           Caribbean  1961.0          55438.0          4.655          66.074   
2           Caribbean  1962.0          56225.0          4.471          66.444   

   ...  accessElectricity  agriculturalLand  agricultureTractors  \
0  ...                NaN               NaN                  NaN   
1  ...                NaN              20.0                  NaN   
2  ...                NaN              20.0                  NaN   

   cerealProduction  fertilizerHa        co2  greenhouseGases     co2_PC  \
0               NaN           NaN  11092.675              NaN  204.62037   
1               NaN           NaN  11576.719              NaN  208.82281   
2               NaN           NaN  12713.489              NaN  226.11808   

   pm2.5_35  battleDeaths  
0       NaN           NaN  
1       NaN           NaN  
2       NaN           NaN  

[3 rows x 25 columns]

(c) Yes, the data looks good

# 6. Wealth

6.1

In [61]:
names = pd.unique(data['name'])
na = data[data["name"].isna()]
na
unique_name = set()
for name in names: 
    unique_name.add(name)

In [62]:
print(data["name"].nunique())

249


There are 249 different countries in the data

6.2

In [95]:
min(data.time)

1960.0

In [96]:
max(data.time)

2019.0

The most recent year is 2019 and earliest year is 1960

6.3

In [104]:
data_gdp = data[~data.GDP_PC.isna()]
max(data_gdp.time)

2019.0

2019 is the most recent year for GDP data

6.4

In [105]:
data_gdp.GDP_PC.mean()

12048.778024852181

The average GDP across all countries as of 2019 is 12048.778024852181

6.5

In [109]:
most_recent = max(data_gdp.time)
country_most_recent = data_gdp.groupby("name")["time"].max()
(most_recent - country_most_recent).nlargest(5)

name
Djibouti                    9.0
Faroe Islands               9.0
Liechtenstein               9.0
Turks and Caicos Islands    9.0
Eritrea                     8.0
Name: time, dtype: float64

Here are 5 countries with the largest number of the most recent years missing

6.6

In [115]:
data_gdp[data_gdp["time"] == 2019].groupby("region")["GDP_PC"].mean().sort_values(ascending=False)

region
Europe      33227.983767
Americas    14516.826303
Asia        14116.124163
Oceania     11817.735554
Africa       2765.708956
Name: GDP_PC, dtype: float64

Yes I think this order is reasonable since Europe and Americas are two biggest economy. Africa is underdeveloped so the GDP level is lower than the rest of the world on average.  

6.7

In [143]:
gdp_2019 = data_gdp[data_gdp["time"] == 2019]
max_2019 = gdp_2019.GDP_PC.idxmax()

rich = gdp_2019.loc[max_2019]
rich[["name", "GDP_PC", "totalPopulation"]]

name                Luxembourg
GDP_PC             111043.5317
totalPopulation       620001.0
Name: 7039, dtype: object

Luxembourg is the richest country

In [144]:
max_2019 = gdp_2019.GDP_PC.idxmin()

poor = gdp_2019.loc[min_2019]
poor[["name", "GDP_PC", "totalPopulation"]]

name                  Burundi
GDP_PC              208.07473
totalPopulation    11530580.0
Name: 843, dtype: object

Burundi is the poorest country 

6.8

In [189]:
continent = data_gdp["region"].dropna().unique().tolist()
result = []
for name in continent: 
    temp = gdp_2019[gdp_2019["region"] == name]
    max_index = temp.GDP_PC.idxmax()
    min_index = temp.GDP_PC.idxmin()
    richest = temp.loc[max_index]
    poorest = temp.loc[min_index]
    rich_info = richest[["region","name", "GDP_PC", "totalPopulation"]]
    poor_info = poorest[["name", "GDP_PC", "totalPopulation"]]
    summary = pd.concat((poor_info, rich_info), axis = 0)
    result.append(summary)


In [191]:
final_result = pd.concat((result[0], result[1]), axis=1)
for i in result[2:]:
    
    final_result= pd.concat((final_result, i),axis=1)

final_result = final_result.transpose()
final_result

name      GDP_PC totalPopulation    region        name  \
0            Haiti   1245.0125      11263077.0  Americas     Bermuda   
1      Afghanistan   573.28755      38041754.0      Asia       Qatar   
0          Burundi   208.07473      11530580.0    Africa  Seychelles   
0          Ukraine  3224.93517      44386203.0    Europe  Luxembourg   
0  Solomon Islands  1749.42726        669823.0   Oceania   Australia   

        GDP_PC totalPopulation  
0  94106.36125         64027.0  
1  63281.67468       2832067.0  
0  15048.74693         97625.0  
0  111043.5317        620001.0  
0  57183.37935      25365745.0

In [192]:
final_result.columns = ["poorest", "GDP", "population", "region","richest", "GDP", "population"]
final_result.set_index("region")

poorest         GDP  population     richest          GDP  \
region                                                                       
Americas            Haiti   1245.0125  11263077.0     Bermuda  94106.36125   
Asia          Afghanistan   573.28755  38041754.0       Qatar  63281.67468   
Africa            Burundi   208.07473  11530580.0  Seychelles  15048.74693   
Europe            Ukraine  3224.93517  44386203.0  Luxembourg  111043.5317   
Oceania   Solomon Islands  1749.42726    669823.0   Australia  57183.37935   

          population  
region                
Americas     64027.0  
Asia       2832067.0  
Africa       97625.0  
Europe      620001.0  
Oceania   25365745.0

6.9

From the chart above, we can tell the difference for the GDP for the poorest and richest countries differs a lot with 100s even 1000s times difference. I didn't expect Bermuda is the richest country in America. I'm not sure why, but I think it might because it's wide spread of international business. It is easier to develop the economy and infrastucture so that's why rich countries tend to have small populations.

# 7 Health

7.1

In [30]:
data_le = data
data_le.head(3)

iso3   name iso2    region                       sub-region  \
0  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
1  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
2  ABW  Aruba   AW  Americas  Latin America and the Caribbean   

  intermediate-region    time  totalPopulation  fertilityRate  lifeExpectancy  \
0           Caribbean  1960.0          54211.0          4.820          65.662   
1           Caribbean  1961.0          55438.0          4.655          66.074   
2           Caribbean  1962.0          56225.0          4.471          66.444   

   ...  accessElectricity  agriculturalLand  agricultureTractors  \
0  ...                NaN               NaN                  NaN   
1  ...                NaN              20.0                  NaN   
2  ...                NaN              20.0                  NaN   

   cerealProduction  fertilizerHa        co2  greenhouseGases     co2_PC  \
0               NaN           NaN  11092.675              NaN  204.62037   
1               NaN           NaN  11576.719              NaN  208.82281   
2               NaN           NaN  12713.489              NaN  226.11808   

   pm2.5_35  battleDeaths  
0       NaN           NaN  
1       NaN           NaN  
2       NaN           NaN  

[3 rows x 25 columns]

In [19]:
data_1960 = data_le[data_le["time"] == 1960]
data_1960.lifeExpectancy.isna().value_counts()

False    189
True      27
Name: lifeExpectancy, dtype: int64

There are 27 countries who don't have lifexpectancy data for 1960

In [20]:
data_2019 = data_le[data_le["time"] == 2019]
data_2019.lifeExpectancy.isna().value_counts()

False    198
True      19
Name: lifeExpectancy, dtype: int64

There are 19 countries who don't have lifexpectancy data for 1960

7.2

In [21]:
data_le.lifeExpectancy.nlargest(1)

10581    85.41707
Name: lifeExpectancy, dtype: float64

In [22]:
data_le.lifeExpectancy.nsmallest(1)

6097    18.907
Name: lifeExpectancy, dtype: float64

In [29]:
data_le.loc[10581][["name", "time"]]

name    San Marino
time        2012.0
Name: 10581, dtype: object

In [28]:
data_le.loc[6097][["name", "time"]]

name    Cambodia
time      1977.0
Name: 6097, dtype: object

- Shortest LE: 18.907; Year: 1977; Country: Cambodia
- Longest LE: 85.41707; Year: 2012; Country: San Marino

7.3

I found out there is an event called "Cambodian Genocide", a systematic persecution and killing of Cambodians by the Khmer Rouge under the leadership of Communist Party of Kampuchea general secretary Pol Pot, who radically pushed Cambodia towards an entirely self-sufficient agrarian socialist society (Wiki, 2022)

Wikimedia Foundation. (2022, January 14). Cambodian genocide. Wikipedia. Retrieved January 16, 2022, from https://en.wikipedia.org/wiki/Cambodian_genocide 

7.4

In [42]:
continent = data_le["region"].dropna().unique().tolist()
result = []
for name in continent: 
    temp = data_le[data_le["region"] == name]
    max_index = temp.lifeExpectancy.idxmax()
    min_index = temp.lifeExpectancy.idxmin()
    longest = temp.loc[max_index]
    shortest = temp.loc[min_index]
    long_info = longest[["region","name", "lifeExpectancy"]]
    short_info = shortest[["name", "lifeExpectancy"]]
    summary = pd.concat((long_info, short_info), axis = 0)
    result.append(summary)

In [43]:
final_result = pd.concat((result[0], result[1]), axis=1)
for i in result[2:]:
    
    final_result= pd.concat((final_result, i),axis=1)

final_result = final_result.transpose()
final_result

region            name lifeExpectancy                    name  \
0  Americas  Cayman Islands       82.19024                   Haiti   
1      Asia       Hong Kong       85.07805                Cambodia   
0    Africa         Algeria          76.88                  Rwanda   
0    Europe      San Marino       85.41707  Bosnia and Herzegovina   
0   Oceania       Australia           82.9        Papua New Guinea   

  lifeExpectancy  
0         41.762  
1         18.907  
0         26.172  
0         60.353  
0         38.935

In [44]:
final_result.columns = ["region", "longest_country", "lifeExpectancy","shortest_country", "lifeExpectancy"]
final_result.set_index("region")

longest_country lifeExpectancy        shortest_country lifeExpectancy
region                                                                        
Americas  Cayman Islands       82.19024                   Haiti         41.762
Asia           Hong Kong       85.07805                Cambodia         18.907
Africa           Algeria          76.88                  Rwanda         26.172
Europe        San Marino       85.41707  Bosnia and Herzegovina         60.353
Oceania        Australia           82.9        Papua New Guinea         38.935

The graph here has the info or country with longest and shortest LE for each continent

7.5

In [55]:
data_good = data_le.dropna(subset=["lifeExpectancy"], axis=0)
index = data_good.groupby("name").time.idxmin()
index

name
Afghanistan                 60
Albania                    182
Algeria                   3251
Angola                     120
Antigua and Barbuda        544
                         ...  
Viet Nam                 12634
Virgin Islands (U.S.)    12574
Yemen                    12815
Zambia                   12935
Zimbabwe                 12995
Name: time, Length: 204, dtype: int64

In [63]:
earliest = data_good.loc[index][["name", "time", "lifeExpectancy"]]
earliest

name    time  lifeExpectancy
60               Afghanistan  1960.0        32.44600
182                  Albania  1960.0        62.28300
3251                 Algeria  1960.0        46.14100
120                   Angola  1960.0        37.52400
544      Antigua and Barbuda  1960.0        61.96800
...                      ...     ...             ...
12634               Viet Nam  1960.0        59.03900
12574  Virgin Islands (U.S.)  1960.0        66.22485
12815                  Yemen  1960.0        29.91900
12935                 Zambia  1960.0        46.68700
12995               Zimbabwe  1960.0        53.01900

[204 rows x 3 columns]

Here is the table for the first LE for each country

7.6

In [64]:
lastest = data_good.loc[data_good.groupby("name").time.idxmax()][["name", "time", "lifeExpectancy"]]
lastest

name    time  lifeExpectancy
119              Afghanistan  2019.0        64.83300
241                  Albania  2019.0        78.57300
3310                 Algeria  2019.0        76.88000
179                   Angola  2019.0        61.14700
603      Antigua and Barbuda  2019.0        77.01600
...                      ...     ...             ...
12693               Viet Nam  2019.0        75.40000
12633  Virgin Islands (U.S.)  2019.0        79.66829
12874                  Yemen  2019.0        66.12500
12994                 Zambia  2019.0        63.88600
13054               Zimbabwe  2019.0        61.49000

[204 rows x 3 columns]

Here is the table for the most recent LE for each country

7.7

In [69]:
merged_data = pd.merge(earliest, lastest, on="name")
merged_data

name  time_x  lifeExpectancy_x  time_y  lifeExpectancy_y
0              Afghanistan  1960.0          32.44600  2019.0          64.83300
1                  Albania  1960.0          62.28300  2019.0          78.57300
2                  Algeria  1960.0          46.14100  2019.0          76.88000
3                   Angola  1960.0          37.52400  2019.0          61.14700
4      Antigua and Barbuda  1960.0          61.96800  2019.0          77.01600
..                     ...     ...               ...     ...               ...
199               Viet Nam  1960.0          59.03900  2019.0          75.40000
200  Virgin Islands (U.S.)  1960.0          66.22485  2019.0          79.66829
201                  Yemen  1960.0          29.91900  2019.0          66.12500
202                 Zambia  1960.0          46.68700  2019.0          63.88600
203               Zimbabwe  1960.0          53.01900  2019.0          61.49000

[204 rows x 5 columns]

In [95]:
n = merged_data.time_y - merged_data.time_x
g = 100 * ((merged_data.lifeExpectancy_y/merged_data.lifeExpectancy_x) ** (1/n) -1)
merged_data["growth_rate"] = g
new_data = merged_data.rename({"time_x":"earliest_year","lifeExpectancy_x":"lifeExpectancy_earliest","time_y":"lastest_year","lifeExpectancy_y":"lifeExpectancy_lastest"},axis=1)
new_data

name  earliest_year  lifeExpectancy_earliest  \
0              Afghanistan         1960.0                 32.44600   
1                  Albania         1960.0                 62.28300   
2                  Algeria         1960.0                 46.14100   
3                   Angola         1960.0                 37.52400   
4      Antigua and Barbuda         1960.0                 61.96800   
..                     ...            ...                      ...   
199               Viet Nam         1960.0                 59.03900   
200  Virgin Islands (U.S.)         1960.0                 66.22485   
201                  Yemen         1960.0                 29.91900   
202                 Zambia         1960.0                 46.68700   
203               Zimbabwe         1960.0                 53.01900   

     lastest_year  lifeExpectancy_lastest  growth_rate  
0          2019.0                64.83300     1.180194  
1          2019.0                78.57300     0.394572  
2          2019.0                76.88000     0.869083  
3          2019.0                61.14700     0.831062  
4          2019.0                77.01600     0.369146  
..            ...                     ...          ...  
199        2019.0                75.40000     0.415452  
200        2019.0                79.66829     0.313738  
201        2019.0                66.12500     1.353232  
202        2019.0                63.88600     0.532999  
203        2019.0                61.49000     0.251543  

[204 rows x 6 columns]

In [97]:
la_index = g.nlargest(10).index #top 10 fastest growers
new_data.loc[la_index]

name  earliest_year  lifeExpectancy_earliest  lastest_year  \
201        Yemen         1960.0                   29.919        2019.0   
113     Maldives         1960.0                   37.343        2019.0   
114         Mali         1960.0                   28.199        2019.0   
20        Bhutan         1960.0                   34.526        2019.0   
183  Timor-Leste         1960.0                   33.729        2019.0   
0    Afghanistan         1960.0                   32.446        2019.0   
128        Nepal         1960.0                   35.583        2019.0   
66        Gambia         1960.0                   32.054        2019.0   
104      Liberia         1960.0                   34.264        2019.0   
187      Tunisia         1960.0                   42.021        2019.0   

     lifeExpectancy_lastest  growth_rate  
201                  66.125     1.353232  
113                  78.921     1.276385  
114                  59.306     1.268013  
20                   71.777     1.248151  
183                  69.495     1.232786  
0                    64.833     1.180194  
128                  70.778     1.172379  
66                   62.050     1.125813  
104                  64.104     1.067371  
187                  76.699     1.025081

In [98]:
sm_index = g.nsmallest(10).index #top 10 slowest growers
new_data.loc[sm_index]

name  earliest_year  lifeExpectancy_earliest  lastest_year  \
116  Marshall Islands         1987.0                 72.14146        2000.0   
33     Cayman Islands         2010.0                 82.19024        2010.0   
158        San Marino         2012.0                 85.41707        2012.0   
139             Palau         1990.0                 69.06927        2005.0   
191           Ukraine         1960.0                 68.29954        2019.0   
101            Latvia         1960.0                 69.78683        2019.0   
26           Bulgaria         1960.0                 69.24756        2019.0   
107         Lithuania         1960.0                 69.84732        2019.0   
163        Seychelles         1980.0                 69.59756        2019.0   
15            Belarus         1960.0                 67.70810        2019.0   

     lifeExpectancy_lastest  growth_rate  
116                65.23902    -0.770640  
33                 82.19024     0.000000  
158                85.41707     0.000000  
139                69.12927     0.005789  
191                71.82732     0.085396  
101                75.23902     0.127581  
26                 74.91463     0.133413  
107                76.13415     0.146183  
163                73.94390     0.155447  
15                 74.22683     0.155918

Here are two tables of 10 fastest and 10 slowest growers

7.8

One pattern that I noticed is that countries with high growth rate, they tend to have lower life expectancy at the beginning of the period. However, countries with low growth rate, they tend to have high life expectancy at the beginning of the period.

# Total Time: 10hrs